pkg/kubernetes/client/context.go

```diff
@@ -88,54 +87,112 @@ func Contexts() ([]string, error) {
 	return strings.Split(buf.String(), "\n"), nil
 }
 
-// ContextFromIP searches the $KUBECONFIG for a context of a cluster that matches the apiServer
-func ContextFromIP(apiServer string) (cluster, context objx.Map, err error) {
-	kubeconfig, err := Kubeconfig()
+// ContextFromIP searches the $KUBECONFIG for a context using a cluster that matches the apiServer
+func ContextFromIP(apiServer string) (*Cluster, *Context, error) {
+	cfg, err := Kubeconfig()
 	if err != nil {
 		return nil, nil, err
 	}
-	cfg := objx.New(kubeconfig)
 
 	// find the correct cluster
-	cluster = find(cfg.Get("clusters").MustMSISlice(), "cluster.server", apiServer)
-	if cluster == nil { // empty map means no result
-		return nil, nil, fmt.Errorf("no cluster that matches the apiServer `%s` was found. Please check your $KUBECONFIG", apiServer)
+	var cluster Cluster
+	clusters, err := tryMSISlice(cfg.Get("clusters"), "clusters")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := find(clusters, "cluster.server", apiServer, &cluster, ErrorNoCluster(apiServer)); err != nil {
+		return nil, nil, err
 	}
 
 	// find a context that uses the cluster
-	context = find(cfg.Get("contexts").MustMSISlice(), "context.cluster", cluster.Get("name").MustStr())
-	if context == nil {
-		return nil, nil, fmt.Errorf("no context that matches the cluster `%s` was found. Please check your $KUBECONFIG", cluster.Get("name").MustStr())
+	var context Context
+	contexts, err := tryMSISlice(cfg.Get("contexts"), "contexts")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := find(contexts, "context.cluster", cluster.Name, &context, ErrorNoContext(cluster.Name)); err != nil {
+		return nil, nil, err
 	}
 
-	return cluster, context, nil
+	return &cluster, &context, nil
 }
 
+// IPFromContext parses $KUBECONFIG, finds the cluster with the given name and
+// returns the cluster's endpoint
 func IPFromContext(name string) (ip string, err error) {
-	kubeconfig, err := Kubeconfig()
+	cfg, err := Kubeconfig()
 	if err != nil {
 		return "", err
 	}
-	cfg := objx.New(kubeconfig)
 
-	// find the context
-	context := find(cfg.Get("contexts").MustMSISlice(), "name", name)
-	if context == nil {
-		return "", fmt.Errorf("no context named `%s` was found. Please check your $KUBECONFIG", name)
+	// find a context with the given name
+	var context Context
+	contexts, err := tryMSISlice(cfg.Get("contexts"), "contexts")
+	if err != nil {
+		return "", err
 	}
 
-	clusterName := context.Get("context.cluster").MustStr()
-	cluster := find(cfg.Get("clusters").MustMSISlice(), "name", clusterName)
-	if cluster == nil { // empty map means no result
-		return "", fmt.Errorf("no cluster named `%s` as required by context `%s` was found. Please check your $KUBECONFIG", clusterName, name)
+	if err := find(contexts, "name", name, &context, ErrorNoContext(name)); err != nil {
```

Comment:
I'm not sure that passing in an error to a function only for it to return it back is the best way to handle this. Consider the following:

```go
if err := find(contexts, "name", name, &context); err == ErrNotFound {
  return "", ErrorNoContext(name)
} else if err != nil {
  return "", err
}
```